### DISCUSSION TITLE
HA docker e nodered

### DISCUSSION THREAD

**Discussion by Gerrino1965** [indomus/forum#1152] [HUB: Home Assistant]
Ciao. Seguendo la guida di Indomus.it ho fatto la migrazione da home assistant venv a docker. successivamente mi sono accorto che nodered (installato mesi fa seguendo una guida indomus) non comunica più con HA. Ho dei nodi con scritto connecting e il quadretto giallo. Come rimediare?

**Comment by snakuzzo** [comment]
Se non scrivi nel dettaglio che hai fatto, come ti si può aiutare?

**Reply by Gerrino1965** [reply]
Con HA e nodered installati seguendo le vostre guide e perfettamente funzionanti, ho fatto la migrazione seguendo la vostra guida e home assistant funziona regolarmente. (Dimenticavo.... Il tutto su rpi con Raspbian bullseye) Successivamente mi sono collegato da pc a nodered e mi sono accorto che i nodi all'inizio dei flow si presentano con quadratino giallo e la scritta connecting

**Reply by snakuzzo** [reply]
A quale guida ti riferisci di preciso?
Riesci a scrivere qui come hai configurato il tutto?

**Reply by xyzmauro** [reply]
Non è che devi solo rifare il token di lunga durata su HA,copiarlo nel nodo è rifare deploy?

**Reply by snakuzzo** [reply]
Se ha riportato la config dir senza escludere la subdir `.storage` non ha necessità di rigenerare il token. Comunque finché non dettaglia bene quello che ha fatto, nessuno può aiutarlo

**Reply by xyzmauro** [reply]
Su HA ci sarà ancora,ma se butti su un container da zero di Nodered e importi solo i flow potrebbe esserselo perso li

**Reply by snakuzzo** [reply]
Se non hai i volumi persistenti, sì...ma da quello che ha scritto non mi pare abbia creato il container di nodered da zero.

**Reply by xyzmauro** [reply]
La guida per l’installazione di Nodered sul sito mi sembra di ricordare fosse come service,cmq vedremo…😅

**Reply by snakuzzo** [reply]
sì non abbiamo fatto alcuna guida su come avviare nodered su Docker...ma permettimi di dire che è talmente banale che nemmeno serve. Ci vogliono 10 secondi di orologio per migrare

**Reply by xyzmauro** [reply]
Ma certo che ci vogliono 10” netti,per quello dicevo che poi il container è nuovo e importando i flow _forse_ perdi il token

**Reply by snakuzzo** [reply]
vediamo che dice...

**Reply by xyzmauro** [reply]
Yes 😉

**Comment by Gerrino1965** [comment]
Ho installato HA seguendo questa guida (e quelle eventualmente richiamate nell'articolo): https://indomus.it/guide/come-installare-e-configurare-home-assistant-su-un-raspberry-pi-gia-in-uso/
HACS con quest'altra: https://indomus.it/guide/installare-hacs-home-assistant-community-store-sul-proprio-hub/
Il collegamento da remoto con DuckDNS e dehydrated con le seguenti: https://indomus.it/guide/collegarsi-da-remoto-a-home-assistant-installato-su-raspberry-raspbian/
https://indomus.it/guide/aggiornare-automaticamente-duckdns-dal-proprio-raspberry/ (seeguendo il metodo classico non docker)
Ho seguito la seguente per il backup: https://indomus.it/guide/effettuare-il-backup-ed-eventuale-ripristino-di-home-assistant-core-su-raspbian/
Nodered con: https://indomus.it/guide/come-installare-e-configurare-node-red-su-raspberry-pi/

Il problema con nodered si è presentato dopo la migrazione a docker e la configurazione di Caddy seguendo: https://indomus.it/guide/come-migrare-home-assistant-core-da-ambiente-venv-a-docker-su-raspberry-pi/ e https://indomus.it/guide/come-installare-e-configurare-caddy-con-docker-su-raspbian-di-raspberry-pi/

**Reply by Gerrino1965** [reply]
Ora HA e la connessione da remoto con Caddy funzionano, nodered no.
Va reinstallato?
O installato sotto docker?

**Reply by snakuzzo** [reply]
Premesso che...
- Non è che sei obbligato ad installare anche Node-Red su Docker
- Per pulizia...sicuramente meglio installarlo su Docker

...se con HA non hai problemi, ma li hai solo su Node-Red, probabilmente il problema sta nella configurazione che utilizzi.
Come hai configurato la parte di collegamento tra Node-Red ed HA ?
Considera che utilizzando un reverse proxy localmente HA risponde su `http://iplocale:8123` e da internet in `https://tuodominio:tuaporta`.
Prima invece anche in locale rispondeva in https (es. `https://iplocale:8123`)
Poi, utilizzando il reverse proxy, usi anche i `trusted_proxies` nella configurazione di HA, quindi HA accetta solo connessioni dagli IP che hai specificato lì ed in questo caso, a meno che tu non stia passando da internet per collegarti ad HA (spero proprio di no :D ), dovrai per forza di cose adeguare la configurazione (nel log di HA questa cosa la vedi se Node-Red tenta di effettuare un collegamento).

Ora a te la scelta...

Se hai intenzione di lasciare tutto com'è, recupera l'IP LAN di Node-Red e mettilo tra i trusted proxy di HA (se è sullo stesso server puoi usare `127.0.0.1`)

Se hai intenzione di passare a Docker, puoi anche impostare l'IP o la subnet Docker.

Fai vedere come hai configurato HA e come hai configurato la connessione verso HA da Node-Red

**Reply by Gerrino1965** [reply]
Gli indirizzi sono già:
external_url: "https://miodominio.duckdns.org:8123"
internal_url: "http://192.168.1.20:8123"

Per la connessione da nodered non ricordo di aver fatto nulla però è raggiungibile da "https://ip_del_raspberry:1880" e non da "http:"

Cioè in configuration.yaml
la sezione http da così:
http:
   use_x_forwarded_for: true
   trusted_proxies: 
      - 172.17.0.0/16
deve diventare:
http:
   use_x_forwarded_for: true
   trusted_proxies: 
      - 172.17.0.0/16
      - 127.0.0.1
 ?

**Reply by snakuzzo** [reply]
1. `external_url` ed `internal_url` non servono a quello che pensi. Non c'entrano nulla con la configurazione per l'accesso verso HA
2. Se non condividi TUTTI i dettagli io posso risponderti solo andando per un'idea... in linea di massima ti direi sì...dovrebbe bastare aggiungere quello se hai fatto tutto correttamente. Poi boh...se non scrivi cosa hai fatto per interfacciare Node-Red con HA, non posso dirti con sicurezza cosa fare

**Reply by Gerrino1965** [reply]
Che io ricordi ho solo installato l'integrazione "Node-RED Companion".
Ora ho provato a disinstallarla, riavviare HA e installarla nuovamente.
Non mi ha chiesto nessun dato/informazione.
Nei log del raspi appare ripetutamente:
Node-RED[415]: 7 Dec 22:30:44 - [info] [server:Home Assistant] Connecting to https://127.0.0.1:8123

**Reply by snakuzzo** [reply]
vedI? provi a collegarti su `https://127.0.0.1:8123` ... non funzionerà mai nella vita.
Come già ti ho detto prima, ora HA risponde in http in locale e in https da internet. Quindi Node-Red al massimo dovrà collegarsi ad `http://127.0.0.1:8123` e, sempre come detto prima, dovrai assicurarti di avere `127.0.0.1` sotto `trusted_proxies:`

**Reply by Gerrino1965** [reply]
Seguendo i consigli ora ho 127.0.0.1 sotto trusted_proxies:
Nella configurazione di nodered (settings.js) ho commentato la sezione https: e impostato "requireHttps: true".
Ho riavviato
Da browser ora mi connetto a nodered con http anzichè con https, ma su nodered il risultato non cambia.

netstat -an |grep 8123 mi da:
tcp        0      0 0.0.0.0:8123            0.0.0.0:*               LISTEN
tcp        0      0 127.0.0.1:47578         127.0.0.1:8123          TIME_WAIT
tcp        0      0 192.168.1.20:8123       192.168.1.19:11088      ESTABLISHED
tcp        0      0 192.168.1.20:8123       192.168.1.19:11087      ESTABLISHED
tcp        0      0 192.168.1.20:8123       192.168.1.19:11005      TIME_WAIT
tcp        0      0 192.168.1.20:8123       192.168.1.19:11086      ESTABLISHED
tcp6       0      0 :::8123                 :::*                    LISTEN
e  netstat -an |grep 1880:
tcp        0      0 0.0.0.0:1880            0.0.0.0:*               LISTEN
tcp        0      0 192.168.1.20:1880       192.168.1.19:11100      ESTABLISHED
tcp        0      0 192.168.1.20:1880       192.168.1.19:11098      ESTABLISHED
tcp        0      0 192.168.1.20:1880       192.168.1.19:11101      ESTABLISHED
tcp        0      0 192.168.1.20:1880       192.168.1.19:11102      ESTABLISHED
tcp        0      0 192.168.1.20:1880       192.168.1.19:11099      ESTABLISHED
tcp        0      0 192.168.1.20:1880       192.168.1.19:11097      ESTABLISHED

**Reply by snakuzzo** [reply]
Scusa...ma tutto questo che c'entra con quello che ho scritto in precedenza ?
Tu hai detto che nel log vedi questo...

`Node-RED[415]: 7 Dec 22:30:44 - [info] [server:Home Assistant] Connecting to https://127.0.0.1:8123`

Quindi il problema è che Node-Red tenta di collegarsi ad HA utilizzando il protocollo https (che sicuramente usavi prima) ancihé http (visto che HA ora ti risponde in http in locale).

**Reply by Gerrino1965** [reply]
Si. Questo accadeva prima del mio intervento sulla configurazione di Node-Red.
Nella configurazione di nodered (settings.js) ho commentato la sezione https:

Quindi, ricapitolando, seguendo il tuo consiglio:
Ho inserito ho 127.0.0.1 sotto trusted_proxies:
Ed eliminato l'uso di https da parte di Node-Red

Ma ancora non andava. Mancava ancora un pezzetto. 
Ieri sera ho provato a generare un nuovo "Token di accesso a lunga vita" su HA e a inserirlo nella configurazione di Node-Red. 

Ora tutto funziona
Mi piacerebbe sapere perchè è stato necessario

**Reply by snakuzzo** [reply]
perché probabilmente non hai salvato la directory `.storage` di HA non portandoti dietro il token

**Reply by Gerrino1965** [reply]
Veramente, come indicato pere all'inizio della vostra guida ho precauzionalmente fatto il backup dell'intera directory  /home/homeassistant/.homeassistant/ e sottocartelle. Effetuata la migrazione, visto che HA sembrava funzionare come prima non ho fatto nessun ripristino del backup.

**Reply by snakuzzo** [reply]
Non so che dirti...i token vengono riportati correttamente sotto `.storage` e questo te lo possiamo garantire...altrimenti con tutte le migrazioni fatte da server a server non mi funzionerebbe più nulla che utilizzi quei token :) ... invece non ho mai dovuto ricrearli.
Al posto tuo comunque, a questo punto non mi farei tutti questi problemi a posteriori

**Reply by Gerrino1965** [reply]
Come dici tu. Grazie
